Add database

KosSpb opened the pull request:
Смена хранилища InMemory на базу данных.

KosSpb committed:
Project Filmorate v0.3.0

feat: реализовать класс MpaRating, описывающий модель данных рейтинга фильмов;
feat: реализовать класс Genre, описывающий модель данных жанра фильмов;
feat: реализовать класс GenreController для обработки эндпойнтов по жанрам;
feat: реализовать класс MpaRatingController для обработки эндпойнтов по рейтингам;
feat: реализовать классы GenreService, MpaRatingService, FriendshipService предназначенные для соответствующей бизнес-логики, а также внести изменения в методы классов FilmService и UserService с учётом внедрения DAO-классов;
feat: реализовать интерфейсы GenreStorage, FriendshipStorage, LikesStorage,
MpaRatingStorage;
feat: создать DAO-реализации для вышеперечисленных интерфейсов, а также новые для интерфейсов FilmStorage и UserStorage;
feat: добавить в application.properties новые настройки для БД H2;
feat: добавить в pom.xml зависимости для работы с JDBC и БД H2;
delete: удалить класс InMemoryUserStorage и InMemoryFilmStorage;
delete: удалить метод Collection<Long> getUsersIds() из интерфейса UserStorage;

KosSpb committed:
Project Filmorate v0.3.1

feat: добавить тесты для DAO-классов;
fix: добавить default значение для поля confirmation_status;

KosSpb committed:
Project Filmorate v0.3.2

style: названия нескольких методов приведены к camelCase;

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -7,76 +7,100 @@
+            log.info("updateFilm - film id not found: {}", film);
+            throw new NotFoundException("Фильма с данным id не существует.");
+        } else {
+            return filmStorage.updateFilm(film).get();
```
Получается 2 раза вызывается filmStorage.update? Для чего?

KosSpb replied:
Исправлено. 

По поводу файла диаграммы. Я не знал как лучше поступить и создал отдельную ветку для промежуточного ТЗ, и в ней добавил диаграмму и ридми. Перезалью их тогда, чтобы в этой ветке были.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -7,76 +7,100 @@
 
     public Film getFilmById(long id) {
-        return filmStorage.getFilmById(id);
+        if (filmStorage.getFilmById(id).isPresent()) {
```
И тут 2 раза вызывается getFilmById? В целом более удобно пользоваться подходом:
Film film = filmStorage.getFilmById().orElseThrow(() -> new NotFoundException());

KosSpb replied:
Исправлено.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -7,76 +7,100 @@
         }
-        return film;
+
+        return filmStorage.getFilmById(id).get();
```
Опять повторный вызов базы

KosSpb replied:
Исправлено.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FriendshipService.java:
```diff
@@ -0,0 +1,107 @@
+            throw new AlreadyExistException("Друг с данным id уже добавлен.");
+        }
+
+        return userStorage.getUserById(id).get();
```
Опять повторный вызов базы?

KosSpb replied:
Исправлено.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,147 @@
+            String sqlQueryInsGenres = "insert into film_genre(film_id, genre_id) " +
+                    "values (?, ?)";
+            for (Genre genre : film.getGenres()) {
+                jdbcTemplate.update(sqlQueryInsGenres, filmId, genre.getId());
```
Запросы к базе в цикле - очень плохо, лучше делать батч операцию

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,147 @@
+                "where film_id = ?";
+        jdbcTemplate.update(sqlQueryDelGenres, film.getId());
+
+        if (film.getGenres().size() > 0) {
```
Дублирование кода с прошлым методом, лучше вынести в отдельный метод

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,147 @@
+                film.getMpa().getId(),
+                film.getId());
+
+        return getFilmById(film.getId());
```
Зачем делается гет запрос?

KosSpb replied:
Убрал. Были проблемы с порядком жанров при десериализации из JSON. Поэтому заново собирал фильм с помощью БД, чтобы было всё корректно. Теперь изменил тип данных у сета в модели фильма, и всё стало десериализовывать как нужно.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,147 @@
+                .duration(resultSet.getInt("duration"))
+                .mpa(new MpaRating(resultSet.getInt("mpa_rating_id"),
+                        resultSet.getString("mpa_value")))
+                .genres(new LinkedHashSet<>(genresOfFilm(resultSet.getLong("film_id"))))
```
Нельзя делать запросы к базе в цикле, это нужно выносить отдельно и получать по списку фильмов или по единичному фильму

KosSpb replied:
Исправлено.

alexache reviewed (changes requested):
Так же не хватает файла диаграммы со схемой базы данных и ссылкой из ридми

KosSpb committed:
Project Filmorate v0.3.3

fix: добавить в ветку add-database ER-диаграмму и заполненный README.md;
fix: убрать повторные обращения к методам возвращающим Optional в сервисном слое;
refactor: убрать в методах класса FilmDbStorage обращение к базе в цикле;
refactor: вынести код сохранения жанров в БД в отдельный метод, чтобы не дублировать код;

alexache commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -22,5 +20,5 @@ public class Film {
     @NonNull
     private MpaRating mpa;
-    private Set<Genre> genres;
+    private LinkedHashSet<Genre> genres = new LinkedHashSet<>();
```
Лучше писать Set у типа поля

KosSpb replied:
Я помню про это правило, но если указать Set, то при десериализации получается фильм с HashSet вместо LinkedHashSet. Я пока не знаю как это настроить иначе, чем тип жёстко задать.
